server/mediaserver.go

```diff
@@ -673,19 +677,60 @@ func (s *LivepeerServer) HandlePush(w http.ResponseWriter, r *http.Request) {
 	}
 
 	// Do the transcoding!
-	err = processSegment(cxn, seg)
+	tr, err := processSegment(cxn, seg)
 	if err != nil {
 		// TODO return error
 		http.Error(w, err.Error(), http.StatusInternalServerError)
 		return
 	}
+	if tr == nil {
+		http.Error(w, "No sessions available", http.StatusServiceUnavailable)
+		return
+	}
 
+	boundary := common.RandName()
+	accept := r.Header.Get("Accept")
+	if accept == "multipart/mixed" {
+		contentType := "multipart/mixed; boundary=" + boundary
+		w.Header().Set("Content-Type", contentType)
+	}
 	w.WriteHeader(http.StatusOK)
+	w.(http.Flusher).Flush()
+	if accept == "multipart/mixed" {
+		mw := multipart.NewWriter(w)
+		for i, url := range tr.url {
+			mw.SetBoundary(boundary)
+			data := tr.data[i]
```

Comment:
So if verification is used, I think it's possible that `tr.data` could reflect data that was *not* accepted for verification, if all retries result in false negatives and a "higher scoring" result was used. Something like this:

1. Segment_O1 transcoded by O1
2. Verify Segment_O1, verification fails with score -0.5
3. Segment_O2 transcoded by O2
4. Verify Segment_O2, verification fails with score -0.1
5. Segment_O3 transcoded by O3
6. Verify Segment_O3, verification fails with score -0.4
7. Max retries hit, verifier returns Segment_O2 as the "highest scoring" result

However, in this case, `tr.data` will be populated with results from O3. This behavior should be confirmed with a new unit test for the verifier, similar to `TestVerifier_HLSInsertion`. 

Now, there are a few ways to fix this:

1. Pass in the `transcodeResults` into the verifier, and reset `tr.data` in addition to `tr.urls` . This is "easy" but I'm a little skeptical of doing so much bookkeeping on state that is canonically available elsewhere, as in the following...

2. Only return URLs from the `transcodeSegment` function - the verifier already handles resetting those URLs as necessary for HLS segment insertion. Within the multipart writer, we can use the URL and `MemorySession.GetData` if it's not an external URL.

In fact, we don't really need to return anything from the `transcodeSegment` function here; we can reconstruct the URL from the profile name (which is already available) and the segment number. This seems like the safest way to ensure consistency and and requires the fewest changes to `transcodeSegment`, which would be preferable.

Reply:
> we can reconstruct the URL from the profile name (which is already available) 

This dependence on URL structure feels very fragile. All the state manipulation in `verify` already very hacky and hard to follow, adding URL manipulation to `HandlePush` instead of just passing data around will make this worse.
I think, after verifier addition, we should change overall data to look like this:
1. `transcodeSegment` downloads all the transcoded segments.
2. passes downloaded data to verifier
3. and only if it returns success, then saves data to the B's OS. Anyway verification is synchronous, requires all the data to be present and transcoding doesn't considered finished until verification is done.

what do you think?

Reply:
>This dependence on URL structure feels very fragile

We construct the URL in a number of other places though, so it's a pretty safe assumption to make. Could even make it a helper if that makes you feel better. Most importantly to me, it minimizes the scope of changes and hence the reasoning about state that we have to do.

> All the state manipulation in verify already very hacky and hard to follow,

I agree that it's not great, but any changes we make would just mean doing the state manipulation in a slightly different place.

Note that I'm not too happy that HLS insertion is synchronous either - the old approach was nicer in terms of making segments immediately available - but the alternative was to maintain two different code paths for HLS insertion which is worse in terms of maintainability. This was discussed in the verification PR.

> passes downloaded data to verifier and only if it returns success, then saves data to the B's OS..

I feel like we need to be careful about this flow, because it's possible for B and O to both have their own OS, and for O to not upload to B's own OS. Currently, the B handles this situation fine. However, with your approach, we'd end up with a situation where:

* O uploads results to O's OS
* B downloads results from O's OS
* B verifies results (presumably incurring additional transfer through the B)
* B uploads results to B's OS, synchronously

This may have some other unintended side effects as well, such as not using the B's external OS for the verifier. Overall it feels like a fraught change.

Anyway, we can continue to discuss this type of refactoring, but it's really not related to the goals of this PR, so let's keep the changes here focused on what needs to happen for multipart return.

Reply:
> We construct the URL in a number of other places though, so it's a pretty safe assumption to make. Could even make it a helper if that makes you feel better.

Removing reliance on URL composition will make me feel better, not adding more of it.

>  it minimizes the scope of changes

This position makes our worse, with every change.

>  minimizes the scope of changes and hence the reasoning about state that we have to do.

It increases reasoning about state need to be done. 

> I agree that it's not great, but any changes we make would just mean doing the state manipulation in a slightly different place

Right now we're doing unnecessary manipulations - putting data into OS storage (though generic interface), taking it out (using casting!), putting it back - it is unnecessary manipulations, we can decrease number of manipulations and make data flow more straightforward and understandable.

> Currently, the B handles this situation fine. 

How?
- Broadcaster download segment from O's OS
- B uploads segment to own storage 
- B passes URLs to verifier
- Verifier rejects, B transcodes segment again
- B downloads new transcoded segment
- B uploads them to own storage, overwriting old ones
- B passes URLs to verifier 
- Verifier wants to approve segments from the previous round, so it returns older URLs to B
- But that URL are same, data in B's own OS is already overwritten

What I'm missing?


> B verifies results (presumably incurring additional transfer through the B)

Presumably OS will be external - like S3, but verifier will be in internal network, so this would be better?

Reply:
Agree the data flow here could be improved. At this point though, it's a high-risk change in my book because it'll be modifying behavior in scenarios unrelated  to this PR. This is where the "reasoning about state" comes from. Minimizing that risk means avoiding changing anything in how we handle state right now.

The most expedient thing here would be to not change anything about how we handle the `transcodeSegment` function. That keeps this PR scoped strictly to multipart return, which doesn't *need* any other changes outside of the HTTP push handler. We can revisit the data flow when it's not squeezed in with an otherwise unrelated feature.

> B uploads them to own storage, overwriting old ones
> What I'm missing?

The upload goes to a different location precisely to avoid this problem. See [1] . Also, if I understand your proposed approach correctly (eg, downloading the data to a local var rather than pushing directly into an OS immediately) then at the it would not be possible to use remote verifiers, because how else would the verifier receive segments [2]?

Anyway - I agree the state manipulation within the verifier is a bit ugly. But it was done that way for a reason: to keep the verification as self-contained as possible without infiltrating the rest of the broadcaster code, and to minimize the impact to the surrounding code [3]. This is also something we should strive for in general, this feature included.

[1] https://github.com/livepeer/go-livepeer/blob/f41ad9982d6254f1cf217c6834b15c0a02451521/server/broadcast.go#L226-L230

[3]  This admittedly only happens under scenarios where we either have a bug, or the orchestrator doesn't respect the B's upload preferences. But defense in depth says these are scenarios we shouldn't discount if possible. In fact, the initial OS implementation had a bug with exactly this problem.

[2] Policy implementation outline: https://github.com/livepeer/go-livepeer/issues/1100

> Presumably OS will be external - like S3, but verifier will be in internal network, so this would be better?

Don't think we can really expect this assumption to always hold. The least surprising thing to do (and most correct one IMO) is this: if external storage is used, then we should use the external storage URL for all segment handling outside of the node.

Reply:
>  Minimizing that risk means avoiding changing anything in how we handle state right now.
> But it was done that way for a reason: to keep the verification as self-contained as possible without infiltrating the rest of the broadcaster code, 

And then I'll add another piece of code which will bit `a bit ugly`. 
And then again. And again. And soon all our code will be `a bit ugly`. Or not a bit.

> The upload goes to a different location precisely to avoid this problem. 

My mistake, I thought about something else which is actually not happening.
I wrote under assumption that B downloads data and then uploads it into own external O.
But that shouldn't happen under normal circumstances, so my other points are irrelevant.